File: 2023/python/day2.py
#!/usr/bin/env python3

from aoc import *

most = {
    "red": 12,
    "green": 13,
    "blue": 14,
}


def part1(input: str):
    sums = 0
    for g, cubes in delimited(input, ": "):
        id = next(comb(g, "\d+", int))[0]
        try:
            for draw in cubes.split("; "):
                for cube in draw.split(", "):
                    n, color = cube.split()
                    if int(n) > most[color]:
                        raise "Nope"
            sums += id
        except:
            pass
    return sums


def part2(input: str):
    sums = 0
    for g, cubes in delimited(input, ": "):
        id = next(comb(g, "\d+", int))[0]
        maxs = {"red": 0, "green": 0, "blue": 0}
        for draw in cubes.split("; "):
            for cube in draw.split(", "):
                n, color = cube.split()
                maxs[color] = max(maxs[color], int(n))
        debug(maxs)
        x = 1
        for y in maxs.values():
            x *= y
        sums += x
    return sums


if __name__ == "__main__":
    main()

# 224

File: 2023/src/day2.rs
use std::cmp::max;

use nom::{
    bytes::complete::{tag, take_until},
    character::complete::{self, alpha0, multispace0},
    combinator::opt,
    multi::{fold_many_m_n, many0},
    sequence::{preceded, tuple},
    IResult,
};

#[derive(Default)]
pub struct StoneGrab {
    red: u32,
    green: u32,
    blue: u32,
}

fn grab(input: &str) -> IResult<&str, StoneGrab> {
    let (input, stones) = fold_many_m_n(
        1,
        100,
        preceded(
            tuple((opt(tag(",")), multispace0)),
            tuple((complete::u32, multispace0, alpha0)),
        ),
        StoneGrab::default,
        |mut stones: StoneGrab, stone| {
            match stone.2 {
                "red" => stones.red = stone.0,
                "green" => stones.green = stone.0,
                "blue" => stones.blue = stone.0,
                color => panic!("Unknown stone color: {}", color),
            }
            stones
        },
    )(input)?;
    let (input, _) = opt(tag(";"))(input)?;
    Ok((input, stones))
}

fn game(input: &str) -> IResult<&str, Vec<StoneGrab>> {
    let (input, _) = take_until(":")(input)?;
    let (input, _) = tag(":")(input)?;
    many0(grab)(input)
}

#[aoc_generator(day2)]
pub fn parse(input: &str) -> Vec<Vec<StoneGrab>> {
    input.lines().map(|x| game(x).unwrap().1).collect()
}

#[aoc(day2, part1)]
pub fn part1(input: &[Vec<StoneGrab>]) -> usize {
    input
        .iter()
        .enumerate()
        .filter_map(|game| {
            if game
                .1
                .iter()
                .all(|grab| grab.red <= 12 && grab.green <= 13 && grab.blue <= 14)
            {
                Some(game.0 + 1)
            } else {
                None
            }
        })
        .sum()
}

#[aoc(day2, part2)]
pub fn part2(input: &[Vec<StoneGrab>]) -> u32 {
    input
        .iter()
        .map(|game| {
            game.iter().fold(StoneGrab::default(), |mut req, grab| {
                req.red = max(req.red, grab.red);
                req.green = max(req.green, grab.green);
                req.blue = max(req.blue, grab.blue);
                req
            })
        })
        .map(|game| game.red * game.green * game.blue)
        .sum()
}
